src/runtime/dict.cpp

```diff
@@ -1027,6 +1027,28 @@ static int dict_compare(BoxedDict* a, BoxedDict* b) noexcept {
     return res;
 }
 
+static PyObject* dict_richcompare(PyObject* v, PyObject* w, int op) noexcept {
+    Box* res;
+    if (!PyDict_Check(v) || !PyDict_Check(w)) {
+        res = Py_NotImplemented;
+    } else if (op == Py_EQ) {
+        res = dictEq((BoxedDict*)v, (BoxedDict*)w);
+    } else if (op == Py_NE) {
+        res = dictNe((BoxedDict*)v, (BoxedDict*)w);
+    } else {
+        /* Py3K warning if comparison isn't == or !=  */
+        if (PyErr_WarnPy3k("dict inequality comparisons not supported "
+                           "in 3.x",
+                           1) < 0) {
+            return NULL;
+        }
+        res = Py_NotImplemented;
+    }
+    Py_INCREF(res);
```

Comment:
This isn't quite right -- dictEq and dictNe return an owned (already-incref'd) object, so this will end up double-increffing them.  I think we should just say `res = incref(Py_NotImplemented)` on the other paths.